### PR TITLE
Creating array of data for usage on frontend (client side validation)

### PR DESCRIPTION
**Pull request by darkdef**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️/❌
| Fixed issues  | #43

### CODE REVIEW

File: src/Rule/Email.php [hunk 1]
```diff
@@ -126,4 +126,28 @@ public function enableIDN(bool $enableIDN): self
...
+    }
+
+    /**
+     * @inheritDoc
+     * @return array
+     */
+    public function getOptions(): array
+    {
+        return array_merge(
+            $this->allowName ? ['allowName' => true] : [],
```
Comment (thenotsoft) [inline src/Rule/Email.php:146]: Why if the bool value does not always output it, but return an empty array?

eg `['checkDNS' => $this->checkDNS]`

Comment (darkdef) [reply src/Rule/Email.php:146]: If the default value for a flag has not been changed and this flag has not been changed, the result is empty.
If it is formed for the default values, the returned array will be too large.
We do not force the user to specify all values when creating a validator.
;-)

File: src/Rule/Url.php [hunk 2]
```diff
@@ -112,6 +121,23 @@ public function schemes(array $schemes): self
...
+    {
+        return 'url';
+    }
+
+    public function getOptions(): array
+    {
+        return array_merge(
+            ['message' => $this->translateMessage($this->message)],
+            $this->enableIDN ? ['enableIDN' => true] : [],
+            $this->schemesIsChanged ? ['validSchemes' => $this->validSchemes] : [],
```
[the remark this answers is not in the document]
Comment (darkdef) [reply src/Rule/Url.php:138]: ok

File: src/Rule/Email.php (hunk 1 above)
Comment (darkdef) [reply src/Rule/Email.php:138]: It's be corrected

**Comment by darkdef** [comment]
How about this resolve:
In class Rule adding method getDefaultOptions for return values by default and add abstract method getOptions (for required to implementation)
```php
    /**
     * Returns rule options.
     * @return array
     */
    abstract public function getOptions(): array ;

    /**
     * Returns default rule options.
     * @return array
     */
    protected function getDefaultOptions(): array
    {
        return [
            'skipOnEmpty' => $this->skipOnEmpty,
            'skipOnError' => $this->skipOnError,
        ];
    }
```
In child classes 
```php
    public function getOptions(): array
    {
        return array_merge(
            ['min' => $this->min],
            ['message' => $this->translateMessage($this->message, ['min' => $this->min])],
            parent::getDefaultOptions()
        );
    }
```

**Comment by thenotsoft** [comment]
I would prefer this option

```
abstract class Rule implements \JsonSerializable {
    ...

    public function jsonSerialize(): array
    {
        return array_merge(
            $this->getDefaultOptions(), 
            $this->getOptions()
        );
    }

    /**
     * Returns rule options.
     * @return array
     */
    abstract public function getOptions(): array ;

    /**
     * Returns default rule options.
     * @return array
     */
    protected function getDefaultOptions(): array
    {
        return [
            'skipOnEmpty' => $this->skipOnEmpty,
            'skipOnError' => $this->skipOnError,
        ];
    }
    ...
}
```

In child class

```
    public function getOptions(): array
    {
        return [
             'min' => $this->min,
             'message' => $this->translateMessage($this->message, ['min' => $this->min])
        ];
    }
```

**Comment by darkdef** [comment]
Well, if be use api with XML?
Need solution is required with the correct result (as an array) from a single function.

**Comment by thenotsoft** [comment]
> Well, if be use api with XML?
> Need solution is required with the correct result (as an array) from a single function.
> And it violates SRP

How does this option interfere with xml? there is no difference, calling `getOptions` or `jsonSerialize` both return an array, only in the second case you can immediately use json_encode without additional effort.

What is the SRP violation?

**Comment by darkdef** [comment]
Imho class Validate must not be linked with JsonSerializable

**Comment by samdark** [comment]
I agree with @darkdef. Validator should not be responsible for serialization of options.

**Comment by xepozz** [comment]
Shouldn't  `getName` and `getOptions` be provided with usage an interface instead of abstract Rule?

**Comment by darkdef** [comment]
> 
> 
> Shouldn't `getName` and `getOptions` be provided with usage an interface instead of abstract Rule?

We need to return private properties from the parent class, so the getOptions implementation is present in the parent class.
When using the interface in such a situation, the child classes will still not need to implement the getOptions method.
--- 
Нам необходимо вернуть приватные свойства из родительского класса, поэтому реализация getOptions присутствует в родительском классе.
При использовании интерфейса в такой ситуации, от дочерних классов, всё равно не потребуется реализация метода getOptions

**Comment by darkdef** [comment]
In last commit:
Swaped parent::getOptions with child options.
Added getOptions to GroupRule.
Added test to Validator on create with incorrect rule

**Comment by xepozz** [comment]
Can you please add a basic definition of the name, for example `static::class`?

**Comment by darkdef** [comment]
> Can you please add a basic definition of the name, for example `static::class`?
And extract className from fully qualified name with namespace?
With usage explode/pcre/reflection ?

See my commits - this was the first implementation ;-)
```php
    public function getName(): string
    {
        return strtolower((new \ReflectionClass($this))->getShortName());
    }
```
But @samdark against. Keep it simple, no magic

**Comment by xepozz** [comment]
Why I should to implement `getName` method if my validator will never be used in the frontend?

**Comment by xepozz** [comment]
> And extract className from fully qualified name with namespace?
With usage explode/pcre/reflection ?

No, just fully qualified name.

File: src/Rule/Number.php [hunk 3]
```diff
@@ -114,4 +114,34 @@ private function getNotANumberMessage(): string
         }
         return 'Value must be a number.';
     }
+
+    /**
+     * @param mixed $value the data value to be checked.
+     */
+    private function isNotNumber($value): bool
```
[the remark this answers is not in the document]
Comment (darkdef) [reply src/Rule/Number.php]: Yes, this unused. Cutted

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!

**Pull request merged**